File: packages/sdk-typescript/src/observability.ts
/**
 * Sprint MAOS-B / P1 — observability helpers.
 *
 * Today: `recordHandoff` — stamp a typed delegation envelope on the
 * outgoing PDP request body so the parent's terminal span carries
 * `handoff_to_did` / `handoff_task` / `handoff_expected_output` /
 * `handoff_rationale`. The dashboard surfaces these as edge labels in
 * the swarm action graph and, in P3, as planned-vs-actual diffs.
 *
 * The handoff is parent-declared — it lives on the request that
 * immediately precedes the fork, *not* on `forkChild()` (which is
 * child-bound env wiring). Keep the two surfaces separate.
 */
export interface SpanHandoffEnvelope {
  /** DID of the child agent receiving the delegation. */
  toAgentDid: string;
  /** What the child is expected to do (1-2 sentences). */
  task: string;
  /** Optional structured output the parent expects back. */
  expectedOutput?: string;
  /** Optional rationale — why this fork, why this child. */
  rationale?: string;
}

/**
 * apiCall body shape the SDK posts to `/v1/proxy/:command`. Permissive
 * by design — we only need to attach `handoff` without dictating the
 * caller's broader shape.
 */
export interface ApiCallLike {
  handoff?: SpanHandoffEnvelope;
  [k: string]: unknown;
}

/**
 * Attach a typed handoff envelope to an outgoing apiCall body. Returns a
 * new object — does not mutate the input (callers commonly reuse the
 * apiCall shape across retries).
 *
 * No PDP-side validation beyond zod length caps. The handoff has zero
 * effect on the authorization decision; it is a pure observability
 * annotation. Caller-supplied `handoff` already on the apiCall wins.
 */
export function recordHandoff<T extends ApiCallLike>(apiCall: T, handoff: SpanHandoffEnvelope): T {
  if (apiCall.handoff) return apiCall;
  return {
    ...apiCall,
    handoff: {
      toAgentDid: handoff.toAgentDid,
      task: handoff.task,
      ...(handoff.expectedOutput !== undefined ? { expectedOutput: handoff.expectedOutput } : {}),
      ...(handoff.rationale !== undefined ? { rationale: handoff.rationale } : {}),
    },
  };
}

/**
 * Sprint MAOS-B / P2 — prompt + reasoning capture envelope.
 *
 * Caller ships the plaintext prompt over TLS. Control-plane gates capture
 * on customer config + ToS + sample-rate, then redacts PII and AEAD-
 * encrypts before insert. PDP forwards as-is; never inspects, never
 * persists. Capture is OFF by default for every customer.
 */
export interface SpanPromptEnvelope {
  text: string;
  reasoning?: string;
}

/**
 * Attach a prompt + (optional) reasoning blob to an outgoing apiCall body.
 * Returns a new object — does not mutate. Caller-supplied `prompt` on the
 * apiCall wins (no clobbering an explicit attachment).
 *
 * Use when the SDK wraps an LLM call and wants to capture the input that
 * drove the tool decision. The control-plane decides whether to actually
 * persist it; the SDK does not need to check config.
 */
export function attachPrompt<T extends ApiCallLike>(apiCall: T, prompt: SpanPromptEnvelope): T {
  if ((apiCall as { prompt?: unknown }).prompt) return apiCall;
  return {
    ...apiCall,
    prompt: {
      text: prompt.text,
      ...(prompt.reasoning !== undefined ? { reasoning: prompt.reasoning } : {}),
    },
  } as T;
}

File: packages/sdk-python/src/nomos/observability.py
"""Sprint MAOS-B / P1 — observability helpers (Python mirror of TS SDK).

``record_handoff`` stamps a typed delegation envelope on the outgoing
``api_call`` dict that the SDK posts to ``/v1/proxy/:command``. PDP
persists it on the parent's terminal span as four flat columns
(``handoff_to_did``, ``handoff_task``, ``handoff_expected_output``,
``handoff_rationale``); the dashboard surfaces it as an edge label in
the swarm action graph and P3 will diff declared-vs-actual handoffs.

Handoff is parent-declared, so it lives on the request that immediately
precedes the fork — NOT on ``fork_child`` (which is child-bound env
wiring). The two surfaces are intentionally separate.
"""

from __future__ import annotations

from typing import Any, Optional, TypedDict


class SpanHandoffEnvelope(TypedDict, total=False):
    """Structured handoff stamped on the outgoing proxy call.

    ``to_agent_did`` and ``task`` are required; the other two fields are
    optional. Length caps are enforced server-side by zod.
    """

    to_agent_did: str
    task: str
    expected_output: str
    rationale: str


def record_handoff(
    api_call: dict[str, Any],
    *,
    to_agent_did: str,
    task: str,
    expected_output: Optional[str] = None,
    rationale: Optional[str] = None,
) -> dict[str, Any]:
    """Return ``api_call`` with a typed ``handoff`` envelope attached.

    Returns a shallow copy — does not mutate the input (callers often
    reuse the api_call across retries). When ``api_call`` already has a
    ``handoff`` key, the caller wins and this function is a no-op.

    Mirrors the TS SDK's ``recordHandoff`` shape. PDP applies no
    authorization side-effects from this field — it is a pure
    observability annotation.
    """
    if "handoff" in api_call and api_call["handoff"]:
        return api_call
    handoff: dict[str, Any] = {
        "toAgentDid": to_agent_did,
        "task": task,
    }
    if expected_output is not None:
        handoff["expectedOutput"] = expected_output
    if rationale is not None:
        handoff["rationale"] = rationale
    return {**api_call, "handoff": handoff}


class SpanPromptEnvelope(TypedDict, total=False):
    """P2 — prompt + reasoning capture envelope.

    Caller ships plaintext over TLS. Control-plane gates capture on
    customer config + ToS + sample-rate, then redacts PII and AEAD-
    encrypts before insert. PDP never inspects, never persists.
    """

    text: str
    reasoning: str


def attach_prompt(
    api_call: dict[str, Any],
    *,
    text: str,
    reasoning: Optional[str] = None,
) -> dict[str, Any]:
    """Attach a prompt + optional reasoning blob to an outgoing api_call.

    Returns a shallow copy — does not mutate. Caller-supplied ``prompt``
    on the api_call wins (no clobbering).
    """
    if "prompt" in api_call and api_call["prompt"]:
        return api_call
    prompt: dict[str, Any] = {"text": text}
    if reasoning is not None:
        prompt["reasoning"] = reasoning
    return {**api_call, "prompt": prompt}


__all__ = [
    "SpanHandoffEnvelope",
    "SpanPromptEnvelope",
    "attach_prompt",
    "record_handoff",
]
